Feature infinite scroll

philippe-cleany opened the pull request.

philippe-cleany committed:
package: add react-infinite-scroller

We add react-infinite-scroller package version 1.0.15.
We do this in order to create a new list component that will use
infinite scroll.

philippe-cleany committed:
list: add InfiniteList component

We add new component: <InfiniteList />.
this compoent render a scrollable list that load data if there still get
data. this component use react-infinite-scroller component.
We do this in order to have scrollable list in the app.

philippe-cleany committed:
reducer: change behavior of list reducer

We allow reducer to know when we use infiniteList or List component,
and the add couple of different action for the reducer.
We to this in order to load more data in case of infiniteList is query
the reducer.

philippe-cleany committed:
action: add infiniteList prop

We add inifiniteList prop in order to no if we use infiniteList or List
for query the right query for the user.

philippe-cleany committed:
infiniList: factorization of the code

We factorizing the code, in order to have less code.

sebastien-cleany commented on src/reducer/resource/list/ids.js:
```diff
@@ -5,8 +5,17 @@ export default resource => (previousState = [], { type, payload, requestPayload,
+          if (requestPayload.pagination.page === 1) {
+            return payload.data.map(record => record.id);
+          }
+          return previousState.concat(payload.data.map(record => record.id));
```
if (meta.infiniteList || requestPayload.pagination.page === 1) {
  return payload.data.map(record => record.id);
}
return previousState.concat(payload.data.map(record => record.id));

philippe-cleany replied:
it is not this behavior that we want

sebastien-cleany commented on src/mui/list/InfiniteList.js:
```diff
@@ -0,0 +1,182 @@
+import defaultTheme from '../defaultTheme';
+import InfiniteScroll from 'react-infinite-scroller';
+import LinearProgress from 'material-ui/LinearProgress';
+import { List } from './List';
```
Is all this include needed ?

philippe-cleany replied:
absolutely !

sebastien-cleany commented on src/mui/list/InfiniteList.js:
```diff
@@ -0,0 +1,182 @@
+};
+
+export class InfiniteList extends List {
+  constructor(props, state) {
```
useless constructor

philippe-cleany replied:
ok

philippe-cleany replied:
ok

sebastien-cleany commented on src/mui/list/InfiniteList.js:
```diff
@@ -0,0 +1,182 @@
+      }
+  }
+
+  getQuery() {
```
const query = super.getQuery()
query.page = this.props.params.page;
return query;

philippe-cleany replied:
ok

sebastien-cleany commented on src/mui/list/InfiniteList.js:
```diff
@@ -0,0 +1,182 @@
+      return query;
+  }
+
+  updateData(query) {
```
Is dup needed ?

philippe-cleany replied:
Yes because we set GrudGetList with diff parameters

sebastien-cleany commented on src/mui/list/InfiniteList.js:
```diff
@@ -0,0 +1,182 @@
+  }
+
+  getNextPage() {
+       this.changeParams({type: 'INC_PAGE'})
```
weird indent

philippe-cleany replied:
ok

sebastien-cleany commented on src/mui/list/InfiniteList.js:
```diff
@@ -0,0 +1,182 @@
+}
+
+const getLocationSearch = props => props.location.search;
+const getQuery = createSelector(
```
Useless ?

philippe-cleany replied:
nope !

sebastien-cleany commented on src/mui/list/InfiniteList.js:
```diff
@@ -0,0 +1,182 @@
+  }
+}
+
+const getLocationSearch = props => props.location.search;
```
useless ?

philippe-cleany replied:
nope !

sebastien-cleany commented on src/mui/list/InfiniteList.js:
```diff
@@ -0,0 +1,182 @@
+);
+
+function mapStateToProps(state, props) {
+    const resourceState = state.admin[props.resource];
```
const resourceState = state.admin[props.resource];
const newProps = ListMapStateToProps(state, props);
newProps.hasMore: resourceState.list.ids.length < resourceState.list.total;
return newProps

sebastien-cleany commented on src/mui/list/InfiniteList.spec.js:
```diff
@@ -0,0 +1,58 @@
+            </InfiniteList>
+        );
+        const textElement = wrapper.find('CardText').children();
+        console.log('%c [InfiniteList]', 'color:red;',textElement.text());
```
console

philippe-cleany committed:
code review: remove console.log

We removing console.log and the constructor

philippe-cleany committed:
code review

We modifyed getQuery method by using the extended method of list
and replacing the diff parameter for infiniteList component.
We do this in order to simplify the code and to have less code de write.

philippe-cleany committed:
code review: fix same key error

In the app when we load data the list item don't have an unique key.
And we have error in the console. Now we create an unique key for the
data that we load.

philippe-cleany committed:
infinitelist: fix same key fetched

With the infiniteList when we reload the page that already been set.
We trying to fetch the same data with the same key. It not a good way
of programing and we have log error in the console. Now we concat the
data with an ES6 method that return the unique key of arrays. We do this
in order to fix that log error and to optimize the app.
